Add details of sort and delete leave in DG (#248)
AY2324S1-CS2103T-W12-1/tp: docs/DeveloperGuide.md

* `SortCommandParser#parse()` — Parses the input arguments by storing the prefixes of its respective values as an `ArgumentMultimap`,
and creates a new `SortCommand` object with the parsed field and order.

The `SortCommand` object then communicates with the `Model` API by calling the following methods:

* `Model#updateSortedEmployeeList(field, order)` — Calls `sortEmployees` method of `AddressBook` class.
* `AddressBook#sortEmployees(field, order)` — Calls `sortEmployees` method of `UniqueEmployeeList` class.
* `UniqueEmployeeList#sortEmployees(field, order)` — Sorts the internal list according to the field and order given.

The following sequence diagram below shows how the sort operation works:

![Sort Sequence Diagram](images/SortSequenceDiagram.png)

Given below is an example usage scenario for the command:

**Step 1**: The user launches the application.

**Step 2**: The user executes the `sort f/FIELD in/ORDER` command in the CLI
(note that only `name`, `salary`, `overtime` and `leaves` fields, as well as `asc` and `desc` orders are valid).

**Step 3**: The list of all employees in the employee book will be sorted accordingly.

> The following activity diagram summarizes what happens when a user executes the reset command:

Reviewer: i think this is supposed to be sort instead of reset?

* `Model#setEmployee(Employee, Employee)` — Sets the employee in the existing employee list to the new `Employee` object which has been edited by `DeleteLeaveCommand#execute()`.
* `Model#updateFilteredEmployeeList(Predicate)` — Updates the view of the application to show all employees.

The method `DeleteLeaveCommand#execute()` returns a `CommandResult` object, which stores information about the completion of the command.

The following sequence diagram below shows how the operation of deleting a leave appointment works:

![Delete Leave Sequence Diagram](images/DeleteLeaveSequenceDiagram.png)

Given below is an example usage scenario for the command:

**Step 1**: The user launches the application.

**Step 2**: The user executes the `deleteleave id/EMPLOYEE_ID from/START_DATE to/END_DATE` command in the CLI.
* `START_DATE` and `END_DATE` are inputs of format `yyyy-MM-dd`.

**Step 3**: Leave appointments that fall between the start and end dates will be deleted from the employee specified with the employee ID input.
* If no leave appointments exist between the start and end dates, an error will be produced and shown to the user.

> The following activity diagram summarizes what happens when a user executes the edit leave command:

Reviewer: i think this is supposed to be delete leave command?

1. Shutdown using the UI

>    1. Click the `X` button in the top right corner of the window.<br>

Reviewer: i think the correct grammatical phrasing is "on the top right corner"?

Author: This was an alert made by IntelliJ, and I double checked on the internet, most people used "in" instead of "on".

   1. Prerequisites: List all persons using the `list` command. At least 1 employee is in the list.

   2. Test case: `sort f/salary in/asc`<br>
   Expected: Employees will be sorted based on their salaries in ascending order. 
   Details of the deleted employee shown in the status message.

   3. Test case: `sort f/phone in/asc` or `sort f/department in/desc`<br>
   Expected: List is not sorted (field phone cannot be used to sort). Error details shown in the status message.

   4. Test case: `sort f/name in/ascending` or `sort f/department in/random`<br>
   Expected: List is not sorted (order parameter can only be asc or desc). Error details shown in the status message.

   5. Test case: `sort f/name` or `sort in/desc`<br>
   Expected: List is not sorted (missing parameters). Error details shown in the status message.
   
   6. Test case: `sort f/name in/` or `sort f/ in/desc`<br>
      Expected: List is not sorted (empty parameters). Error details shown in the status message.

2. Sorting the list of employees when only some employees are being shown

   1. Prerequisites: Filter some employees using the `find Marketing` command. Some employees in the list.

>    2. Try the test cases in the previous section (Adding leave while all employees are being shown)

Reviewer: I think you may have forgotten to chaneg "Adding leave" to "Sorting employees"

Author: Yes thanks! I must have missed many wrong terms when I copypasted the structure from another command.

   1. Test case: `report`<br>
      Expected: Invalid command format error message is shown in the status bar.

### Sorting a list

1. Sorting the list of employees while all employees are being shown

   1. Prerequisites: List all employees using the `list` command. At least 1 employee is in the list.

   2. Test case: `sort f/salary in/asc`<br>
   Expected: Employees will be sorted based on their salaries in ascending order. 

>    Details of the deleted employee shown in the status message.

Reviewer 2: I think this needs to be changed?

   3. Test case: `sort f/phone in/asc` or `sort f/department in/desc`<br>
   Expected: List is not sorted (field phone cannot be used to sort). Error details shown in the status message.

   4. Test case: `sort f/name in/ascending` or `sort f/department in/random`<br>
   Expected: List is not sorted (order parameter can only be asc or desc). Error details shown in the status message.

   5. Test case: `sort f/name` or `sort in/desc`<br>
   Expected: List is not sorted (missing parameters). Error details shown in the status message.
   
   6. Test case: `sort f/name in/` or `sort f/ in/desc`<br>
      Expected: List is not sorted (empty parameters). Error details shown in the status message.

2. Sorting the list of employees when only some employees are being shown

   1. Prerequisites: Filter some employees using the `find Marketing` command. Some employees in the list.

   2. Try the test cases in the previous section (Sorting the list of employees while all employees are being shown)

>       Expected: Same as the previous section

Reviewer 2: perhaps it would be good to state that only the filtered employee list will be sorted?

Author: I think that is because it's pertaining to the previous section, not the current section.

      Expected: No employee leave is deleted (ID does not exist). Error details shown in the result display.

   4. Test case: `deleteleave id/EID12345678 from/2023-12-04 to/2023-12-05` or `deleteleave id/EID1234-5678 from/2023-30-11 to/2023-30-11`<br>
      Expected: No employee leave is deleted (incorrect field format). Error details shown in the result display.

   5. Test case: `deleteleave id/ from/2023-12-04 to/2023-12-05` or `deleteleave id/EID1234-5678 from/ to/2023-12-05` or `deleteleave id/EID1234-5678 from/2023-12-04 to/`<br>
      Expected: No employee leave is deleted (empty fields). Error details shown in the result display.

   6. Test case: `deleteleave from/2023-12-04 to/2023-12-05` or `deleteleave id/EID1234-5678 to/2023-12-05` or `deleteleave id/EID1234-5678 from/2023-12-04`<br>
      Expected: No employee leave is deleted (missing parameters). Error details shown in the result display.

   7. Test case: `deleteleave id/EID1234-5678 from/2023-12-05 to/2023-12-04`<br>
      Expected: No employee leave is deleted (invalid date order). Error details shown in the result display.

   8. Test case: `deleteleave id/EID1234-5678 from/2023-12-04 to/2023-12-05`<br>

>       Expected: No employee leave is deleted (no leaves exist between "2023-12-03" and "2023-12-05"). Error details shown in the result display.

Reviewer 2: ```suggestion
      Expected: No employee leave is deleted (no leaves exist between "2023-12-04" and "2023-12-05"). Error details shown in the result display.
```

* **Documentation**

   Being new to writing user and developer guides or creating UML diagrams, we had to spend a significant amount of time looking at existing docs and learning to write our own. We also had to spend a lot of time ensuring that the documentation was consistent with the code and our implementation. This was a challenge because it required an in-depth understanding of the codebase, and so we had to defer most of the technical parts of our documentation towards the end of the project.

### Effort Required

### Achievements of the Project

## Planned Enhancements

### Remark not shown in GUI employee list

Problem:
* In our current implementation, the remark is not displayed on the employee list in GUI. 
* Users have to use report command to display employee remarks in message box.

Solution:

> * We will update the employee card in the GUI to also display remarks of the employees.

Reviewer: I'm not sure if this a good solution though the employee card might become too cluttered... i think maybe can have a feature that allows you to generate report upon selecting the employee card instead?

### Deleting a remark requires typing the whole remark

Problem:
* In our current implementation, deleting a remark requires the user to type in the remark exactly as it is
  typed in when added, i.e. exactly as it is shown in the message box.
* This is highly inefficient and time-consuming.

Solution:
* A solution is to address a remark by its index (based on date added) instead of its value, so that
  `deleteremark id/EMPLOYEE_ID 1` is enough to delete the first remark of employee with
  employee id EMPLOYEE_ID.

### Adding a leave with a date far in the past and future

Problem:
* In our current implementation, users are able to add a leave for an employee with date that is too far in the past
  and in the future, such as `1111-11-11` and `2222-12-12`.
* While users should be able to add a past and future dates as leaves for record keeping purposes,
  we see that being able to input a date this extreme is excessive and highly unrealistic.

Solution:
* We will enhance the leave list so that there are three separate lists, one for current year, one for last year, and one for next year.

> * Users will only be allowed to add leaves from last year to next year, and only the list for the current year will be considered when counting for leaves.

Reviewer: maybe can phrase "when counting for leaves" to "counting number of annual leaves allocated"?

### Two employees are allowed to have the same phone number or email

Problem:
* In our current implementation, adding an employee with the same phone number or email as another existing employee is allowed.
* This is caused by our decision to only check same employee by comparing by employee id.
* This might lead to unintended duplicates.

Solution:
* We will update the check for the same employee to check not only by employee id but also by phone number and email.

### Unable to view the number of leaves remaining when adding a leave period that exceeds maximum number of leaves

Problem:
* In our current implementation, if a user tries to add a leave for an employee but that exceeds the maximum number of allowed employee leaves,
  the app only displays an error message telling the user it exceeds the remaining number of leaves.
* However, the app does not show the remaining number of leaves, so the fastest method is to use report command to display the current amount of
  leaves taken by the employee and perform manual subtraction to obtain remaining number of leaves.

Solution:

> * We plan to revamp the error messages pertaining to this issue to display as much information as possible to the user, 

Reviewer: can maybe give an example of the new error message?

  in this case the remaining number of leaves.

### Unclear error message for invalid email address

Problem:
* In our current implementation, when a user keys in an invalid email address, the error message shown is very long and slightly inconsistent.
* This will highly confuse users who have very complex email addresses and keys the wrong email address by mistake.

Solution:

> * We intend to keep error messages as short but as unambiguous as possible, so that users are able to identify

Reviewer: might be a little too vague? can maybe give an example of the new error message?

Another feature we are proud of is our `report` feature that allows our users to generate reports as txt files for each employee. While it might not contain a lot of information yet, we believe that it was a great starting point considering the time-constraint we had for this project and we plan to extend and further develop this feature, improving the experience for our users.

Overall, we are proud of our project and we believe that we have done our best with all the constraints and challenges we faced. We are happy with the result and believe that our product will meet the needs of our target audience.

## Planned Enhancements

### Remark not shown in GUI employee list

Problem:
* In our current implementation, the remark is not displayed on the employee list in GUI. 
* Users have to use report command to display employee remarks in message box.

Solution:
* We will update the employee card in the GUI to display remarks of the employees upon clicking on it.

> * This decision is to ensure the displayed employee card is not too cluttered with too much information.

Reviewer 2: ```suggestion
* This is to ensure that the displayed employee card is not cluttered with too much information.
```

### Deleting a remark requires typing the whole remark

Problem:
* In our current implementation, deleting a remark requires the user to type in the remark exactly as it is
  typed in when added, i.e. exactly as it is shown in the message box.

> * This is highly inefficient and time-consuming.

Reviewer 2: Instead of saying highly inefficient, perhaps it can replace be with something like "this can be tedious, especially for long remarks"

Solution:
* A solution is to address a remark by its index (based on date added) instead of its value, so that
  `deleteremark id/EMPLOYEE_ID 1` is enough to delete the first remark of employee with
  employee id EMPLOYEE_ID.

### Adding a leave with a date far in the past and future

Problem:
* In our current implementation, users are able to add a leave for an employee with date that is too far in the past
  and in the future, such as `1111-11-11` and `2222-12-12`.
* While users should be able to add a past and future dates as leaves for record keeping purposes,
  we see that being able to input a date this extreme is excessive and highly unrealistic.

Solution:

> * We will enhance the leave list so that there are three separate lists, one for current year, one for last year, and one for next year.
> * Users will only be allowed to add leaves from last year to next year, and only the list for the current year will be
>   considered when counting the number of annual leaves allocated.

Reviewer 2: ```suggestion
* We will enhance the leave list so that there are three separate lists, one for the current year, one for the previous year, and one for the following year.
* Users will only be allowed to add leave dates that fall within the time period of the beginning of the previous year to the end of the next year, and the number of annual leaves allocated will be calculated accordingly.
* For example, users will be able to add leave dates that fall in 2022, 2023, and 2024. Leaves taken in each year will be tracked separately.
```
